### PR TITLE
[TTPReq] Allow trying another payment method after PIN required error

### LINKED ISSUE #14171: [TTPReq] Fallback payment method and 'Offline PIN' handling
Designs: pdfdoF-5KF-p2
Flows: pdfdoF-5GF-p2

## Description

To meet fallback payment requirements, we can extend the payment failure modal to include an option for selecting another payment method. This would close the failure modal and show the checkout view. [Human Interface Guidelines](https://href.li/?https://developer.apple.com/design/human-interface-guidelines/tap-to-pay-on-iphone#Displaying-results) allow reusing the checkout screen for displaying fallback payment options.

## Solution

1. After TTP payment fails with `offline_pin_required` error, add an additional button to `Try Another Payment Method`
2. For `offline_pin` errors, make sure the error message indicates the reason for the failure

## More Details

To properly support fallback payment methods in all markets (especially UK) we need to fulfill these requirements:

1. All merchants must be set up with at least one (1) accepted alternative payment method so that all
transactions can be completed in person
2. Ensure the app provides a seamless transition between the Tap to Pay on iPhone UI and the accepted
alternative payment method provided for the merchant. 

## Design and User Flows

| Existing Modal | Flow |
| --- | --- |
| Image 1 | Image 2 |

### PR DESCRIPTION
**Pull request by staskus**
Closes: #14171

## Description

Some cards are not able to complete contactless transactions using a PIN, which can result in payment failure. In regions where a significant number of cards cannot be used with contactless under specific circumstances (e.g. Offline PIN markets) the app must provide alternative payment methods and ensure a seamless transition between TTP UI and the alternative payment method.

## Solution

I made the changes to handle the PIN error and dismiss the modal with "Try Another Payment Method" option rather than returning to the order. Limiting these changes for Tap to Pay since the requirement only concerns this payment method.

Although the fix is relatively straightforward there are multiple places where this logic could be introduced. There are two main parts:
1. Handling the dismiss action: whether the app navigates back to the order or stays on the payment method selection. This logic belongs to the place that presents the payment UI in the first place. Therefore, I decided `PaymentMethodViewModel` holds the navigation part of the logic.
2. Showing "Try another payment method" rather than "Back to order" on modals. This logic belongs only to specific IPP flow, therefore only IPP models should be affected. 

### Error Handling Logic:
* Implemented a new method `errorRequiresFallbackPaymentMethod` in `BuiltInCardReaderPaymentAlertsProvider` to check if an error requires a fallback payment method.
* Added the `requiresFallbackPaymentMethod` property to the `CardPaymentErrorProtocol` and its conforming types to determine if a fallback payment method is needed based on the error type. [[1]](diffhunk://#diff-0c3adf508810c6f3b2f211cb853971249def207885516de3c148ae660c8c1c38R980) [[2]](diffhunk://#diff-0c3adf508810c6f3b2f211cb853971249def207885516de3c148ae660c8c1c38R1003-R1012) [[3]](diffhunk://#diff-0c3adf508810c6f3b2f211cb853971249def207885516de3c148ae660c8c1c38R1036-R1044)

### Additions to error modals:
* Added a new parameter `requiresFallbackPaymentMethod` to various modal error classes (`CardPresentModalError`, `CardPresentModalErrorEmailSent`, `CardPresentModalErrorWithoutEmail`, `CardPresentModalNonRetryableError`, `CardPresentModalNonRetryableErrorEmailSent`, and `CardPresentModalNonRetryableErrorWithoutEmail`) to determine if a fallback payment method is required. [[1]](diffhunk://#diff-99ea148b9844cf6d7a5ef130bdb81655ce7289f52251b057c90043fd4b51c246R43-R52) [[2]](diffhunk://#diff-f90ed76bc5a5061daa5350ce5646ee76efc331ae19c72afb1e5fe49b8c79ba9aR43-R51) [[3]](diffhunk://#diff-f146bf7c9708025df6078d137a423f29a10ecb6a23e42875c94df055c8eefcb1R40-R48) [[4]](diffhunk://#diff-0bba9870c6e76d205eb6c3ad8bfc4a18717c1cdbe84f16d6dc6289782a95342fR47-R53) [[5]](diffhunk://#diff-42efa4db281defc7a3ebc50b1dfa8be48064cded98dae08bd16a48bf2c58f8c4R48-R53) [[6]](diffhunk://#diff-8a7ed1e13df685ba9d754d069f0517da0a8e1c01923c68ea2e5cf3483a84e2c0R43-R48)
* Updated the `Localization` extension to provide different button labels based on the `requiresFallbackPaymentMethod` parameter. [[1]](diffhunk://#diff-99ea148b9844cf6d7a5ef130bdb81655ce7289f52251b057c90043fd4b51c246L103-L116) [[2]](diffhunk://#diff-0bba9870c6e76d205eb6c3ad8bfc4a18717c1cdbe84f16d6dc6289782a95342fL89-R97)

### Modal UI fixes:
* Modified the `primaryButton`, `secondaryButton`, and `auxiliaryButton` styles to make font size adjustment work by setting `lineBreakMode` to `.byClipping`. [[1]](diffhunk://#diff-9cab1b19c4132b588a65c18367cf12e058b47df70ec5ec302e4eeaaa6dfafe71R191-R198) [[2]](diffhunk://#diff-9cab1b19c4132b588a65c18367cf12e058b47df70ec5ec302e4eeaaa6dfafe71R207)

### Payment Methods Handling:
* Updated the `PaymentMethodsViewModel` to handle cases where a fallback payment method is required by not calling the `onFailure` callback immediately.

## Steps to reproduce

Prerequisites:
- An iPhone and store that supports TTP

### Offline PIN error

1. Make a 1.02 payment
2. Get an error
3. Confirm "Try Another Payment Method" button is shown instead of "Back to Order"
4. Tap "Try Another Payment Method"
5. Confirm the modal is dismissed

### Offline PIN error + Retry

1. Make a 1.02 payment and get an error
2. Tap "Try Collecting Again"
3. Confirm "Try Another Payment Method" button is shown
4. Tap "Try Another Payment Method"
5. Confirm the modal is dismissed

### Offline PIN error + email options

1. Return `return onCompletion(true)` in `isEligibleSendingReceiptAfterPayment` `ReceiptEligibilityUseCase`
1. Make a 1.02 payment and get an error
3. Confirm "Try Another Payment Method" button is shown along "Email Receipt Button"
4. Tap "Try Another Payment Method"
5. Confirm the modal is dismissed

### Other error

1. Make a 1.05 payment and get an error
3. Confirm "Back to Order" button is shown
5. Tap "Back to Order"
6. Confirm the modal is dismissed and the view navigated back to the order

## Testing information

- Tested cases using iPhone 14 Pro 17.7
- Tested with US and UK stores. UK shows a PIN entry for the online-PIN option (1.03 amount)
- Added Payment method view model tests

### PIN required errors

| PIN Error | Retried PIN Error | PIN Error + Email | Retried PIN Error + Email |
|-----------|-------------------|-------------------|---------------------------|
| IMG_9580 | IMG_9581 | IMG_9578 | IMG_9579 |

### Other errors

| Error | Retried Error | Error + Email | Retried Error + Email |
|-------|---------------|--------------|-----------------------|
| IMG_9582 | IMG_9584 | IMG_9585 | IMG_9586 |

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift [hunk 1]
```diff
@@ -268,7 +268,13 @@ final class PaymentMethodsViewModel: ObservableObject {
                 // Update order in case its status and/or other details are updated after a failed in-person payment
                 self?.updateOrderAsynchronously()
 
-                onFailure()
+                switch error {
+                case let error as CardPaymentErrorProtocol where error.requiresFallbackPaymentMethod:
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift]: Just wondering: Should we put this switch-case filtering into a separate function? 🤔 Something like `shouldFallbackToOtherPaymentMethods(for discoveryMethod: CardReaderDiscoveryMethod, error: Error) -> Bool`

File: WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift [hunk 2]
```diff
@@ -100,21 +102,43 @@ extension CardPresentModalError {
             }
         }
 
-        static func noThanks(transactionType: CardPresentTransactionType) -> String {
+        static func dismiss(transactionType: CardPresentTransactionType, requiresFallbackPaymentMethod: Bool) -> String {
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalError.swift:105]: Good rename here, the `noThanks` made me chuckle thought 😂

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift [hunk 3]
```diff
@@ -188,12 +188,14 @@ private extension CardPresentPaymentsModalViewController {
         primaryButton.applyPrimaryButtonStyle()
         primaryButton.titleLabel?.adjustsFontSizeToFitWidth = true
         primaryButton.titleLabel?.minimumScaleFactor = 0.5
+        primaryButton.titleLabel?.lineBreakMode = .byClipping
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift:191]: TIL `.byClipping`. I definitely had trouble with moving to the new ButtonConfiguration API.

**Review by iamgabrielma — APPROVED** [review]
Works great, tested on iPhone 14 iOS 17.7.1

✅ Offline PIN error
✅ Offline PIN error + Retry
✅ Offline PIN error + email options
✅ Other error

Question: What's the use case for presenting the email receipt button if the payment fails? (It shows as well as in the "Other error" case).

File: WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift (hunk 1 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift]: Good idea 👍

**Pull request merged**